File: problems/678. Valid Parenthesis String/solution.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        low = hi = 0
        for c in s:
            low += 1 if c == "(" else -1
            hi += 1 if c != ")" else -1
            if hi < 0:
                break
            low = max(low, 0)
        return low == 0

File: problems/678. Valid Parenthesis String/solution.cpp
class Solution
{
public:
    bool checkValidString(string s)
    {
        stack<int> open, stars;
        for (int i = 0; i < s.size(); i++)
        {
            switch (s[i])
            {
            case '(':
                open.push(i);
                break;
            case '*':
                stars.push(i);
                break;
            default:
                if (open.empty() && stars.empty())
                    return false;
                if (open.empty())
                    stars.pop();
                else
                    open.pop();
            }
        }
        while (!open.empty())
        {
            if (stars.empty() || open.top() > stars.top())
                return false;
            stars.pop();
            open.pop();
        }

        return true;
    }
};
